packages/postgres-database/src/simcore_postgres_database/utils_tags.py

```diff
@@ -0,0 +1,243 @@
+""" Repository pattern, errors and data structures for models.tags
+"""
+
+import functools
+from dataclasses import dataclass
+from typing import Any, Optional, TypedDict
+
+import sqlalchemy as sa
+from aiopg.sa.connection import SAConnection
+from simcore_postgres_database.models.groups import user_to_groups
+from simcore_postgres_database.models.tags import tags, tags_to_groups
+from simcore_postgres_database.models.users import users
+
+
+#
+# Errors
+#
+class BaseTagError(Exception):
+    pass
+
+
+class ValidationError(BaseTagError):
+    pass
+
+
+class TagNotFoundError(BaseTagError):
+    pass
+
+
+class TagOperationNotAllowed(BaseTagError):  # maps to AccessForbidden
+    pass
+
+
+#
+# Repository: interface layer over pg database
+#
+class TagDict(TypedDict, total=True):
+    # NOTE: ONLY used as returned value, otherwise used
+    id: int
+    name: int
+    description: str
+    color: str
+
+
+_TAG_COLUMNS = [tags.c.id, tags.c.name, tags.c.description, tags.c.color]
+
+
+@dataclass(frozen=True)
+class TagsRepo:
+    user_id: int
+
+    @classmethod
+    def _validate_data(
```

Comment:
Q: what made you choose dataclass over pydantic in this case? why creating validating code?

Reply:
If there was  a reason for the above, please add a note in the code.

Reply:
Great question. Let me clarify the rationale here. 
- I use ``dataclass`` because I have developed this class in ``simcore_postgres_database``
- This is an interface on top of the raw aiopg/sqlalchemy interface
- There will be another interface on top of this with pydantic models (next PRs) 
- On second thought, I think that the validation indeed belongs there and not here. This layer should have the guarantees from the later above. I will change it.